pennylane/ops/qubit.py

```diff
@@ -1975,15 +1976,19 @@ def __init__(self, *params, control_wires=None, wires=None, control_values=None,
         # etc. The positioning of the block is controlled by padding the block diagonal
         # to the left and right with the correct amount of identity blocks.
 
-        padding_left = control_int * len(U)
-        padding_right = 2 ** len(wires) - len(U) - padding_left
+        self._padding_left = control_int * len(U)
+        self._padding_right = 2 ** len(wires) - len(U) - self._padding_left
+        self._CU = None
 
-        CU = block_diag(np.eye(padding_left), U, np.eye(padding_right))
+        super().__init__(*params, wires=wires, do_queue=do_queue)
 
-        params = list(params)
-        params[0] = CU
+    def _matrix(self, *params):
+        if self._CU is None:
+            self._CU = block_diag(np.eye(self._padding_left), self.U, np.eye(self._padding_right))
 
-        super().__init__(*params, wires=wires, do_queue=do_queue)
+        params = list(params)
+        params[0] = self._CU
+        return super()._matrix(*params)
```

Comment:
Here we are changing `ControlledQubitUnitary` so that the full controlled matrix is not loaded upon initialization but instead when the `matrix` attribute is used. Caching is used, so that the matrix only need be constructed once.

The motivation for doing this is that `MultiControlledX` (which inherits from `ControlledQubitUnitary`) could be limited in size due to memory requirements because (without this change) `ControlledQubitUnitary` immediately loads the matrix, even if it's not needed because the op will be decomposed.

In other words, with this change we're able to create much bigger `MultiControlledX` gates and defer the simulation to the device (which will either use the matrix or the decomp).